### PR TITLE
added undeclared params validator #810

### PR DESCRIPTION
**Pull request by hobofan**
Only allow declared parameters as proposed in #810.

Works like this:

```
params do
  optional :beer
  optional :wine
  optional :juice
  declared_only
end
```

Now if there is a parameter present that is not `beer`, `wine` or `juice` a validation exception will be thrown.
Also tested for nested parameters.

It is not quite finished yet:
- I plan on adding a `except` option to skip the validator for some parameters. Especially in combination with the bug described in #1050 / #1085 , a param that is accepted on every endpoint would have to either be declared in every params block or will throw an exception with this validator
- Not tested/tried it for Arrays yet.
- Some feedback on the name of `declared_only` would be nice. I not sure if it fits in well.

### CODE REVIEW

**Comment by dblock** [comment]
I like it. I think that the DSL could be clearer and be on `params`, maybe turn it around, for example `params undeclared: :error do ...`. I think that if a param is accepted with every endpoint it should be in a `helpers ...` and reused via `use`, introducing `except` will defeat the purpose of this IMO. 

I think I would like to be able to turn this globally on for my entire API. 

I'd like a way to turn this on gradually, we have 400 endpoints in a project and we would love to make params declared only, so maybe something like `params undeclared: :warn`?

**Comment by dblock** [comment]
Something that's not declared is smuggled? Should `undeclared` be `smuggled`? j/k

**Comment by hobofan** [comment]
Great, I would also prefer to have it on `params`, but decided to follow the conventions of the existing validator (even though it works somewhat different than the other validators).

**Comment by hobofan** [comment]
@dblock 

> I'd like a way to turn this on gradually, we have 400 endpoints in a project and we would love to make params declared only, so maybe something like params undeclared: :warn?

I've put it in there now, but it isn't easily testable and I haven't really found any similar behaviour in grape yet.

**Comment by dblock** [comment]
I really like this. I think it needs README and CHANGELOG and it would be good to go. Testing via expecting a warning is good AFAIK.

**Comment by hobofan** [comment]
I had to change a bit more, to get the validation for generic hashes working in an expected way.

E.g.:
```ruby
params undeclared: :error do
  optional :config, type: Hash
end
```

If you only specify the type Hash, you probably want to allow any Hash.

In contrast, if you declared you Hash with a block like
```ruby
params undeclared: :error do
  optional :config, type: Hash do
    requires :language
  end
end
```
you have a strict expectation of how the Hash should look like.

As for the warning test, I tried my best (even ugly stuff like expect_any_instance_of) but didn't manage to get a correct expectation running.

**Comment by hobofan** [comment]
When trying to integrate it into our own API I noticed that it failed when a Hash had more than one nested parameter defined.
This is now fixed but requires more involved Hash inclusion checking for the declared params. On the other hand this opens up the possibility to expand `given` blocks to work with nested blocks like:

```ruby
params do
  optional :a, type: Hash do
    optional :b
  end
  given :a => :b do
    requires :c
  end
end
```

File: lib/grape/dsl/parameters.rb [hunk 1]
```diff
@@ -163,9 +163,38 @@ def given(attr, &block)
...
+        convert_hash = lambda do |obj|
+          return [obj, true] unless obj.is_a?(Array) || obj.is_a?(Hash)
+          if obj.is_a? Array
+            return obj.map { |e| convert_hash.call(e) }
+          elsif obj.is_a? Hash
+            obj.map do |k, v|
+              return k, convert_hash.call(v).to_h
+            end
+          end
+        end
```
Comment (dblock) [inline lib/grape/dsl/parameters.rb]: Should be refactored out of here.

File: lib/grape/dsl/parameters.rb [hunk 2]
```diff
@@ -163,9 +163,38 @@ def given(attr, &block)
...
+          key = smaller.keys.first
+
+          return included.call(bigger[key], smaller[key])
+        end
+
+        conv_declared_params = convert_hash.call(@declared_params).to_h
+        conv_param = convert_hash.call([param]).to_h
+
+        included.call(conv_declared_params, conv_param)
+      end
```
Comment (dblock) [inline lib/grape/dsl/parameters.rb]: What if `declared_param?` took a path (an array) instead? Would it make anything simpler?

**Comment by dblock** [comment]
Maybe there's a simpler implementation here that calculates a hash difference (only caring about keys), I haven't dug into things like https://github.com/liufengyun/hashdiff or others, but it sounds like a good way to abstract the problem, no?

**Comment by hobofan** [comment]
Yes, having `declared_param?` accept a path as array of keys might make things simpler. I'll look into it, but might be a week until I have the time.

**Comment by wrtsprt** [comment]
What is the status on this? Is there anything we can do to help? 

This seems to be a quite important feature to me. Especially with being able to turn that on by default ('strict mode') for the whole API.

**Comment by hobofan** [comment]
I sadly won't have time to work on it in the forseeable future, sorry.

I have no problem if anyone picks up where I left off and submits a new PR.

**Comment by dblock** [comment]
I am still very interested in this, it's a super useful feature. Was thinking maybe we should have `strict: true` on `params` as a more rigid DSL, `undeclared: ` still feels odd to me. Looking for someone to finish this PR!

**Comment by hobofan** [comment]
Closing this, as I'm currently clearing out old/outdated PRs of mine.

**Pull request closed without merging**